### PR TITLE
Create an extensive test suite

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

In most cases our test suite is working, but it's (a) hard to read and (b) not complete.

## 🚫 Blocked by

- #774 

## 🔍 What does this change?

- Add an extensive test suite in the `test/` directory with a very high test coverage
- Add a UI test suite to check for expected compilation errors:
    - Ensure, that a `must-use` warning is emitted when using `Report`
    - Ensure the macros cannot be called with wrong parameters (also improved error message on the first parameter of `ensure` if it's not a boolean)
- Fix bugs found by the new test suite:
    - `Report::change_context` did not work for `Report`s created from `eyre` or `anyhow`
    - getting the backtraces from `eyre` and `anyhow` didn't work as expected
    - Remove wrong `Clone` bound on `Stream::attach_printable_lazy`
    - Export the stream extension
    - Remove unneeded bounds on `Report::current_context` (duplicated)
    - Fix `Display` output for `Report`

## 🐾 Next steps

- Improve test run performance by using `nextest`
- Run the doc-tests on `nightly` and with `--all-features` only.
- Check for code coverage in CI

## 🛡 What tests cover this?

🙃 🙂 🙃 🙂

## ❓ How to test this?

Run the test suite locally by running
```
cargo make test
```

## 📹 Demo

cc @hashintel/backend: These tests may be helpful to get more familiar with `error stack`.

|     | **Function Coverage** | **Line Coverage** | **Region Coverage** |
|--------|-------------------|------------------|------------------|
| **Totals** | 95.30% (142/149)  | 93.44% (926/991) | 86.09% (297/345) |

Breakdown:
| **Filename**        | **Function Coverage** | **Line Coverage** | **Region Coverage** |
|---------------------|-----------------------|-------------------|---------------------|
| *compat/anyhow.rs*    | 100.00% (7/7)         | 98.18% (54/55)    | 96.15% (25/26)      |
| *compat/eyre.rs*      | 100.00% (7/7)         | 98.00% (49/50)    | 96.15% (25/26)      |
| *context.rs*          | 100.00% (5/5)         | 100.00% (20/20)   | 88.89% (8/9)        |
| *ext/future.rs*       | 100.00% (10/10)       | 97.73% (86/88)    | 87.50% (14/16)      |
| *ext/iter.rs*         | 100.00% (11/11)       | 100.00% (88/88)   | 100.00% (17/17)     |
| *ext/result.rs*       | 100.00% (8/8)         | 89.23% (58/65)    | 75.86% (22/29)      |
| *ext/stream.rs*       | 100.00% (13/13)       | 98.17% (107/109)  | 91.67% (22/24)      |
| *frame/frame_impl.rs* | 100.00% (21/21)       | 100.00% (102/102) | 100.00% (21/21)     |
| *frame/mod.rs*        | 100.00% (13/13)       | 100.00% (65/65)   | 100.00% (19/19)     |
| *hook.rs*             | 90.00% (9/10)         | 97.37% (37/38)    | 91.67% (11/12)      |
| *iter.rs*             | 73.33% (11/15)        | 73.13% (49/67)    | 79.31% (23/29)      |
| *lib.rs*              | 100.00% (1/1)         | 100.00% (1/1)     | 100.00% (1/1)       |
| *macros.rs*           | 100.00% (4/4)         | 100.00% (12/12)   | 100.00% (4/4)       |
| *report.rs*           | 91.67% (22/24)        | 85.71% (198/231)  | 75.89% (85/112)     |

### CODE REVIEW

File: packages/libs/error-stack/tests/compiletest.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#[cfg_attr(not(nightly), ignore = "Outputs are different on toolchains")]
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/tests/compiletest.rs:1]: I don't think this ignore message quite makes sense. Do you mean different _across_ toolchains or something? (as in between different toolchains)

File: packages/libs/error-stack/tests/test_compatibility.rs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    ONCE.call_once(|| {
+        eyre::set_hook(Box::new(eyre::DefaultHandler::default_with)).expect("Could not set hook");
+    })
+}
+
+#[test]
+#[cfg(feature = "eyre")]
+#[cfg_attr(
+    miri,
+    ignore = "bug: miri is failing for `eyre`, this is unrelated to our implementation"
```
Comment (Alfred-Mountfield) [inline packages/libs/error-stack/tests/test_compatibility.rs:136]: Do we have a tracking issue we can reference here on `eyre`?

**Review by Alfred-Mountfield — COMMENTED** [review]
Few things found (appreciate this is in draft atm)

File: packages/libs/error-stack/tests/compiletest.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/libs/error-stack/tests/compiletest.rs]: Yes, it's *across* toolchains: 8a97f9b7006fbbb83cf834306785a69312f3239d

**Review by Alfred-Mountfield — APPROVED** [review]
Great!

**Pull request merged**